### DISCUSSION TITLE
AsyncClient cookie not being persisted

### DISCUSSION THREAD

**Discussion by cp-weng** [encode/httpx#2144] [Potential Issue]
``` py
# app.py
@router.post("/login")
async def login(response: Response):
    # ...omit
    response.set_cookie(
        key=ACCESS_TOKEN_COOKIE_KEY,
        value=token_value,
    )
    return ...

# conftest.py
@pytest.fixture()
async def async_client(app: FastAPI) -> AsyncGenerator:
    async with AsyncClient(app=app, base_url="http://test") as ac:
        yield ac

# test_blablabla.py
async def test_login(async_client):
    response = await async_client.post("/login", json={...})
    # `response.cookies` is empty here and not being persisted
```

## Expected

As the [docs](https://www.python-httpx.org/quickstart/#cookies) says, `response.cookies` should contain access token

## Actual

`response.cookies` is empty

**Comment by ChihweiLHBird** [comment]
I encountered a similar issue. Wondering if there is any update.

In `response.headers`, I got `Headers({'set-cookie': 'test=pass; Path=/; expires=Sun, 12-Jun-2022 20:13:19 GMT', 'content-length': '4', 'connection': 'keep-alive', 'content-type': 'text/plain; charset=utf-8'})`. But `response.cookies.jar` and `response.cookies` are empty, `<CookieJar[]>`, `<Cookies[]>`

**Reply by cp-weng** [reply]
Here is my current workaround.

```py
import httpx

def get_cookies(set_cookie_header: str) -> httpx.Cookies:
    cookies = httpx.Cookies()
    entries = set_cookie_header.split(", ")
    for entry in entries:
        chunks = entry.split("; ")
        chunk = next(c for c in chunks)
        k, v = chunk.split("=")
        cookies.set(k, v)
    return cookies

async def some_test_case(async_client):
    response = await async_client.post("/login", json={...}) # set cookies in the response
    cookies = get_cookies(response.headers["set-cookie"])
    response = await async_client.post("/consume/cookies", cookies=cookies, json={...}) # pass cookies manually
```

**Reply by lovelydinosaur** [reply]
I can't replicate this, so let's work it through...

I'd always suggest trying to reduce the issue down to the absolute simplest possible case first.
For instance, try starting with this and let me know what you get...

```python
import httpx

client = httpx.Client()
response = client.get("https://httpbin.org/cookies/set?a=b")

print(response)
print(response.headers)
print(response.cookies)
print(client.cookies)
print(client.cookies.get('a'))
```

(You can switch up to async later to check if you get the same results, but start simple)

**Reply by ChihweiLHBird** [reply]
For some other reasons, the client doesn't work in the code I work on. But the async client does not have the cookies as well.

```
method = "get"
url = "some url"

async with self.get_new_session(**session_kwargs) as session:
    try:
        response = await getattr(session, method.lower())(
            url, *args, **kwargs
        )
        print(session.cookies)
        print(session.cookies.get('test'))
        print(response.headers)
        print(response.cookies.jar)
        print(response.text)
```

Result:

```
<Cookies[]>
None
Headers({'set-cookie': 'test=pass; Path=/; expires=Tue, 14-Jun-2022 22:19:55 GMT', 'content-length': '4', 'connection': 'keep-alive', 'content-type': 'text/plain; charset=utf-8'})
<CookieJar[]>
pass
```

https://github.com/sanic-org/sanic-testing/blob/04b1b9672ba18726283697bae84e20bf0a8b684d/sanic_testing/testing.py#L103

**Reply by ahopkins** [reply]
@tomchristie This is for the sanic test client. We've had some flaky tests for a while now where sometimes the cookies test pass, sometimes not. It feels like a race condition, but I'm not sure where. I'll put some time on it this coming week and see if I can narrow down a location. Will report back.

**Comment by lijie97** [comment]
I have the same question
```
Headers({'content-length': '158', 'content-type': 'application/json', 'set-cookie': 'access_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI4aGE3MXh5ZTJ4d3cyeWQ5YnY2YiIsImV4cCI6MTY5NDQ5MDc3MX0.ddYYFuvkWRElAh3IUfbDCvlKzeEinL-wzrlPsYu4Rek; HttpOnly; Max-Age=604800.0; Path=/; SameSite=lax'})
```

The HttpResponse contains a cookie but response.cookie and client.cookie is empty.

**Comment by zombeer** [comment]
Same here, empty `client.cookies` after response with some `set-cookie` headers.
Speaking of asyncclient only.

**Reply by Dogrtt** [reply]
Yep, still the case httpx 0.28.1 AsyncClient
